src/Razor/src/Microsoft.VisualStudio.LanguageServer.ContainedLanguage/DefaultLSPDocumentSynchronizer.cs

```diff
@@ -20,26 +22,95 @@ namespace Microsoft.VisualStudio.LanguageServer.ContainedLanguage
     internal class DefaultLSPDocumentSynchronizer : LSPDocumentSynchronizer
     {
         // Internal for testing
+        private readonly LSPDocumentManager _documentManager;
         internal TimeSpan _synchronizationTimeout = TimeSpan.FromSeconds(2);
         private readonly Dictionary<Uri, DocumentContext> _virtualDocumentContexts;
         private readonly object _documentContextLock = new();
         private readonly FileUriProvider _fileUriProvider;
 
         [ImportingConstructor]
-        public DefaultLSPDocumentSynchronizer(FileUriProvider fileUriProvider)
+        public DefaultLSPDocumentSynchronizer(FileUriProvider fileUriProvider,
+            LSPDocumentManager documentManager)
         {
             if (fileUriProvider is null)
             {
                 throw new ArgumentNullException(nameof(fileUriProvider));
             }
 
+            if (documentManager is null)
+            {
+                throw new ArgumentNullException(nameof(documentManager));
+            }
+
             _fileUriProvider = fileUriProvider;
             _virtualDocumentContexts = new Dictionary<Uri, DocumentContext>();
+            _documentManager = documentManager;
+        }
+
+        internal record SynchronizedResult<TVirtualDocumentSnapshot>(bool Synchronized, TVirtualDocumentSnapshot VirtualSnapshot)
+            where TVirtualDocumentSnapshot : VirtualDocumentSnapshot
+        {
         }
 
+        public override Task<SynchronizedResult<TVirtualDocumentSnapshot>> TrySynchronizeVirtualDocumentAsync<TVirtualDocumentSnapshot>(
+            int requiredHostDocumentVersion,
+            Uri hostDocumentUri,
+            Uri virtualDocumentUri,
```

Comment:
Unfortunately I couldn't figure out a way to create the virtualDocumentUri from within this class due to layering of projects. I don't think it's TOO bad to do it this way, but if someone has a way to avoid it that'd be awesome.